---
title: "compute_twist_metrics"
output: html_document
date: "2023-05-24"
---

```{r setup, include=FALSE}
library(dplyr)
library(ggplot2)
library(tidyverse)
library(readr)
```

# Question

The WES samples were sequenced with the Twist Biosciences Exome 2.0 sequencing panel. We sought to characterise the performance
of the exome sequencing panel.

# Method

For each WES sample, Picard `CollectHsMetrics` (v2.27.5) was used to compute the % on-target bases (1-PCT_OFF_BAIT), 
fold 80 base penalty score (FOLD_80_BASE_PENALTY),
percentage of zero coverage targets (ZERO_CVG_TARGETS_PCT), % of bases in duplicate reads (PCT_EXC_DUPE), 
median read coverage per sample (MEDIAN_TARGET_COVERAGE), % target bases sequenced to a specified coverage (PCT_TARGET_BASES_10X etc.), AT/GC dropout values (AT_DROPOUT, GC_DROPOUT).

We computed the normalised read coverage to assess the extent of GC bias in read coverage. More formally, let $B_j$ be the set of target regions that fall within the $j^th$ GC bin. Then,
$|B_j|$ is the number of target regions within the $j^th$ bin.  Let $N$ be the total number 
of GC bins (in our study, $N = 10$). Also, let $M$ be the total number of target regions captured by the exome sequencing kit. 

Then, let $R_{i}$ be the number of reads that overlap with the $i^{th}$ target region (for $i = 1,2\ldots,M$).  We define the normalised read coverage, labelled `norm_cov`, as  

$norm\_cov = \sum_{j=1}^{N} \sum_{i \in B_{j}} \frac{R_{i}}{L_{i}} \cdot \frac{1}{|B_{j}|}$

# Analysis

## Processing picard `CollectHsMetrics` output
```{r}
picard_hs_metrics <- list.files(path = "results_wes_2_samples_full/", recursive = TRUE, pattern = "output_hs_metrics.txt", full.names = TRUE)

hs_metrics_output <- c()
for (metrics_file_path in picard_hs_metrics){
  # Read the file
  data <- read_lines(metrics_file_path)
  sample_name <-str_extract(metrics_file_path, "(?<=markduplicates\\/)[^\\/]+")
  
  # Extract header information. 66 columns
  header <- strsplit(data[grep("^BAIT_SET", data)], "\t")[[1]]  %>%
    head(-3) # Remove the sample, library and read_group entries
  
  # Extract data
  rows <- data[grep("^hg38_exome_v2", data)]
  
  # Process the data
  processed_data <- strsplit(rows, "\\s+") %>%
    map_dfr(~as.data.frame(t(.), stringsAsFactors = FALSE))
  processed_data
  # Set column names
  colnames(processed_data) <- header
  
  # Convert columns to appropriate types
  processed_data <- processed_data %>%
    mutate(across(.fns = as.numeric, .cols = contains("PCT") | contains("FOLD") | contains("MEAN") | contains("MEDIAN") | contains("TARGET") | contains("PENALTY") | contains("DROPOUT")))
  
  # Append the sample name as the last column
  processed_data$SAMPLE <- sample_name
  processed_data$PCT_ON_BAIT <- 1-processed_data$PCT_OFF_BAIT
  hs_metrics_output <- rbind(hs_metrics_output, processed_data)
}

(hs_metrics_output)
hs_metrics_output_tidy <- hs_metrics_output %>% 
  dplyr::select("PCT_ON_BAIT", "FOLD_80_BASE_PENALTY", "ZERO_CVG_TARGETS_PCT", "SAMPLE", 
                "PCT_EXC_DUPE", "MEDIAN_TARGET_COVERAGE",
                "AT_DROPOUT", "GC_DROPOUT") %>%
  pivot_longer(cols = -SAMPLE, names_to = "metrics", values_to = "values") 

hs_metrics_output_pct_targets <- hs_metrics_output %>% 
  dplyr::select("PCT_TARGET_BASES_10X","PCT_TARGET_BASES_20X","PCT_TARGET_BASES_50X","PCT_TARGET_BASES_100X", "SAMPLE") %>%
  pivot_longer(cols = -SAMPLE, names_to = "metrics", values_to = "values")

```

## Plotting
```{r}
pct_on_bait_plot <- hs_metrics_output_tidy %>%
  dplyr::filter(metrics=="PCT_ON_BAIT") %>%
  ggplot(aes(x=SAMPLE, y=values*100, fill=SAMPLE)) +
  geom_col(position = "dodge")+
  geom_text(aes(label = paste0(round(values*100,1),"%")), position = position_dodge(width = 0.9), hjust = -0.01) +
  theme_minimal()+
  ylab(label="% on-target")+
  scale_fill_viridis_d() + 
  coord_flip(ylim = c(0, 100))

fold_80_plot <- hs_metrics_output_tidy %>%
  dplyr::filter(metrics=="FOLD_80_BASE_PENALTY") %>%
  ggplot(aes(x=SAMPLE, y=values, fill=SAMPLE)) +
  geom_col(position = "dodge")+
  geom_text(aes(label = round(values,3)), position = position_dodge(width = 0.9), hjust = -0.01) +
  theme_minimal()+
  ylab(label="Uniformity (Fold-80 base penalty)")+
  scale_fill_viridis_d() + 
  coord_flip(ylim = c(0, 2.0)) +
  guides(fill="none")

zero_cvg_plot <- hs_metrics_output_tidy %>%
  dplyr::filter(metrics=="ZERO_CVG_TARGETS_PCT") %>%
  ggplot(aes(x=SAMPLE, y=values, fill=SAMPLE)) +
  geom_col(position = "dodge")+
  geom_text(aes(label =paste0(round(values*100,1),"%")), position = position_dodge(width = 0.9), hjust = -0.01) +
  theme_minimal()+
  ylab(label="% zero coverage targets")+
  scale_fill_viridis_d() + 
  coord_flip(ylim = c(0, 0.1))+
  guides(fill="none")

pct_dup_plot <- hs_metrics_output_tidy %>%
  dplyr::filter(metrics=="PCT_EXC_DUPE") %>%
  ggplot(aes(x=SAMPLE, y=values*100, fill=SAMPLE)) +
  geom_col(position = "dodge")+
  geom_text(aes(label = paste0(round(values*100,1),"%")), position = position_dodge(width = 0.9), hjust = -0.01) +
  theme_minimal()+
  ylab(label="% bases in duplicate reads")+
  scale_fill_viridis_d() + 
  coord_flip(ylim = c(0, 10))

pct_median_plot <- hs_metrics_output_tidy %>%
  dplyr::filter(metrics=="MEDIAN_TARGET_COVERAGE") %>%
  ggplot(aes(x=SAMPLE, y=values, fill=SAMPLE)) +
  geom_col(position = "dodge")+
  geom_text(aes(label = paste0(round(values,3),"x")), position = position_dodge(width = 0.9), hjust = -0.01) +
  theme_minimal()+
  ylab(label="Median read coverage in target regions")+
  scale_fill_viridis_d() +
  coord_flip(ylim = c(0, 150))

at_gc_dropout <- hs_metrics_output_tidy %>%
  dplyr::filter(metrics %in% c("AT_DROPOUT", "GC_DROPOUT")) %>%
  tidyr::pivot_wider(names_from = metrics, values_from=values ) %>%
  ggplot(aes(x=AT_DROPOUT, y=GC_DROPOUT, color=SAMPLE)) +
  geom_point()+
  theme_minimal()+
  scale_color_viridis_d() +
  coord_cartesian(xlim=c(0,0.25), ylim=c(0,5))

hs_metrics_output_tidy %>%
  dplyr::filter(metrics %in% c("AT_DROPOUT", "GC_DROPOUT")) %>%
  tidyr::pivot_wider(names_from = metrics, values_from=values )

hs_metrics_output_pct_targets$metrics <- factor(hs_metrics_output_pct_targets$metrics,
                                                levels=c("PCT_TARGET_BASES_10X",
                                                         "PCT_TARGET_BASES_20X",
                                                         "PCT_TARGET_BASES_50X",
                                                         "PCT_TARGET_BASES_100X"))
pct_target_plot <- hs_metrics_output_pct_targets %>%
  ggplot(aes(x=metrics, y=values*100, fill=SAMPLE)) +
  geom_col(position = "dodge")+
  theme_minimal()+
  ylab(label="% target bases at various coverages")+
  scale_fill_viridis_d()+
  guides(fill="none")+
  theme(axis.text.x = element_text(size=6))

png("r_analysis/compute_twist_metrics.png", width = 12, 
    height = 10, units = "in", res = 300)
cowplot::plot_grid(pct_on_bait_plot, fold_80_plot, pct_dup_plot, 
                   pct_median_plot, zero_cvg_plot, at_gc_dropout, pct_target_plot, nrow = 3, labels = "AUTO")
dev.off()
```

We can compare the AT_DROPOUT and GC_DROPOUT plot with figure 6 of Zhou et al., (2021), where the authors compared other commercial panels with the Twist Biosciences `Human Core Exome` panel. In the methods section,
the authors also computed these dropout statistics using the `CollectHsMetrics` tool from Picard. Our results for the newest twist exome 2.0 panel is the opposite of the data shown for the Twist Biosciences `Human Core Exome` panel, as the exome 2.0 panel has a higher GC dropout and lower AT dropout.

```{r}
knitr::include_graphics("r_analysis/zhou2021_fig6.png")
```
Calculate % GC bias in bam file
```{r}
library(GenomicAlignments)
library(GenomicRanges)
library(rtracklayer)
library(BSgenome.Hsapiens.UCSC.hg38)
library(Biostrings)

# Recursively list the paths to all bam files within results dir
bam_file_paths <- list.files(path = "/media/gedac/kane/projects/booney_wes/results_wes_2_samples_full/preprocessing/", pattern = "\\.bam$", full.names = TRUE, recursive=TRUE)
# Load BED file
bed <- import.bed(bed_file)

# Initialise empty df.
combined_df <- data.frame()

for (bam_file_path in bam_file_paths){

  # Extract the term using regular expressions
  sample_name <- str_extract(bam_file_path, "(?<=/)[^/]+(?=/[^/]+\\.md\\.bam)")
  
  # Print the extracted terms
  bed_file <- "hg38_exome_v2.0.2_targets_sorted_validated.re_annotated.sorted.bed"
  
  cat("Loading bam file for: ",sample_name)
  # Load BAM file as pairs
  bam <- readGAlignments(
      file = bam_file_path,
      param = ScanBamParam(
        mapqFilter = 1,
        flag = scanBamFlag(
          isPaired = TRUE,
          isProperPair = TRUE),
        what = c("mapq", "isize")))
  
  # Convert to GAlignment object where range = start(bam@first):end(bam@last)
  bam_granges <- GRanges(bam)
  
  
  # Counts # overlaps of the bam file in each bed file region  (Depth/exome target region)
  overlap_counts <- countOverlaps(bed, bam, type="any")
  
  # Compute the GC content over each interval (span of paired-end reads) in the GRanges object
  freqs <- alphabetFrequency(getSeq(BSgenome.Hsapiens.UCSC.hg38, bed))
  
  getSeq(BSgenome.Hsapiens.UCSC.hg38, bed)
  # This gives the %GC in each exome target region
  gc <- (freqs[,'C'] + freqs[,'G'])/rowSums(freqs)
  
  # I find that for regions with 0% and 100% GC, these are regions with <=10 bases long.
  outliers <- getSeq(BSgenome.Hsapiens.UCSC.hg38, bed[gc<=0.1| gc>=0.9])
  max(width(outliers))
  
  # Create df of gc content for each bed region & depth of coverage
  per_bin_gc_depth <- data.frame(sample_name = sample_name,
    region = bed$name,
                                 region_width = width(bed),
    read_depth = overlap_counts,
                         pert_region_gc = gc) %>%
    dplyr::mutate(norm_read_depth = read_depth / region_width)
  
  combined_df <- rbind(combined_df, per_bin_gc_depth)
  cat(sample_name, "analysis is done!")
}
```

We analyse the GC bias of the twist exome 2.0 exome sequencing panel on a single sample `28RT_CM+`. 

Plot A shows the distribution of the GC content of the twist 2.0 target regions. For each GC bin,
we sum the number of target regions falling within the specified bin. This graph ignores the
# reads that fall within each bin / coverage per bin. Here, we find that whilst
the # exome target regions falling within the central GC bins appear normally distributed,
there are a large number of target regions falling within the extreme %GC bins ([0,0.05] and [0.95,1]). In
plot C, we calculate the mean coverage per GC bin, without normalising for length or number of target regions for 
each bin. The dotted red line shows the median of the mean coverages across all GC bins.

Within these extreme GC bins, in plot B, we find that the length of the target regions are very small, with most regions being SNPs (length 1)
and the maximum width of target regions being 10 bp long.

After removing these extreme GC bins, we first calculate the length-normalised read depth per target region, defined 
to be the ratio of the number of reads overlapping with the target region and the length of the target region. Next,
we sum the length-normalised read depths across all target regions falling within a particular bin. Finally, 
we divide these length-normalised read depths by the number of target regions within each GC bin to account for the uneven 
distribution of target regions at various GC %. We call this value the `normalised read coverage`. The result is shown
in plot E, which shows that there is a bias in coverage towards the GC-poor regions in the samples. The dotted red line shows the median normalised coverage across all GC bins.

To decide which metric best captures the "coverage" across bins, we must ask ourselves 
whether the unnormalised mean coverage, length-normalised coverage or length and #region normalised coverage
make the most sense to interpret the data.


```{r}
# Define the bin ranges
bin_start <- 0
bin_end <- 1
bin_increment <- 0.05

# Extract a single sample for non-cov plots
twoeightRT_CM_per_bin_gc_depth <- combined_df %>%
  dplyr::filter(sample_name == "28RT_CM+")

gc_bin_num_regions <- twoeightRT_CM_per_bin_gc_depth %>%
  dplyr::mutate(bin = cut(pert_region_gc, breaks = seq(bin_start, bin_end, by = bin_increment), include.lowest = TRUE)) %>%
  group_by(bin) %>%
  summarise(bin_sums = n()) %>%
  ggplot(aes(x=bin, y=bin_sums, fill=bin)) +
  geom_col() +
  theme_minimal()+
  theme(axis.text.x=element_text(angle=45), legend.position="none") +
  labs(x="%GC bin", y="# exome target regions")

 gc_bin_log10_width <- twoeightRT_CM_per_bin_gc_depth %>%
  dplyr::mutate(bin = cut(pert_region_gc, breaks = seq(bin_start, bin_end, by = bin_increment), include.lowest = TRUE)) %>%
  group_by(bin) %>%
  ggplot(aes(x=bin, y=log(region_width,base=10), fill=bin)) +
  geom_boxplot(alpha = 0.5) +
  #geom_point(position = position_jitter(), alpha = 0.1)+
  theme_minimal()+
  theme(axis.text.x=element_text(angle=45),legend.position = "none") +
  labs(x="%GC bin", y="Log base-10 of target region width") +
  ylim(c(0,5))
 
 gc_bin_norm_coverage <- combined_df %>%
  dplyr::filter(!(pert_region_gc==0 | pert_region_gc==1)) %>%
  dplyr::group_by(sample_name)%>% 
  dplyr::mutate(bin = cut(pert_region_gc, breaks = seq(bin_start, bin_end, by = bin_increment), include.lowest = TRUE)) %>%
  group_by(sample_name, bin) %>%
  summarise(sample_name = sample_name,
    sum_norm_read_depth = sum(norm_read_depth),
            bin_sums = n()) %>%
   dplyr::distinct() %>%
   dplyr::mutate(norm_read_depth_by_bin_sums = sum_norm_read_depth/bin_sums)
 
 gc_bin_length_coverage_plot <- gc_bin_norm_coverage  %>%
  ggplot(aes(x=bin, y=sum_norm_read_depth)) +
  geom_boxplot()+
  geom_point(aes(color=sample_name), position=position_jitter(width=0.2))+
  geom_hline(aes(yintercept=median(norm_read_depth_by_bin_sums)),linewidth=1, linetype = 'dotted', col = 'red')+
  theme_minimal()+
  scale_color_viridis_d() +
  theme(axis.text.x=element_text(angle=45),legend.position="none") +
  labs(x="%GC bin", y="Length-normalised coverage per bin")
 
 gc_bin_norm_coverage
gc_bin_norm_coverage_plot <- gc_bin_norm_coverage  %>%
  ggplot(aes(x=bin, y=norm_read_depth_by_bin_sums)) +
  geom_boxplot()+
  geom_point(aes(color=sample_name), position=position_jitter(width=0.2))+
  geom_hline(aes(yintercept=median(norm_read_depth_by_bin_sums)),linewidth=1, linetype = 'dotted', col = 'red')+
  theme_minimal()+
  scale_color_viridis_d() +
  theme(axis.text.x=element_text(angle=45),legend.position="none") +
  labs(x="%GC bin", y="Normalised coverage per bin")

gc_bin_mean_coverage <- combined_df %>%
  dplyr::group_by(sample_name)%>% 
  dplyr::mutate(bin = cut(pert_region_gc, breaks = seq(bin_start, bin_end, by = bin_increment), include.lowest = TRUE)) %>%
  group_by(sample_name, bin) %>%
  summarise(mean_coverage = mean(read_depth)) %>%
  dplyr::distinct()

gc_bin_mean_coverage_plot <- gc_bin_mean_coverage %>%
  ggplot(aes(x=bin, y=mean_coverage)) +
  geom_boxplot()+
  geom_point(aes(color=sample_name), position=position_jitter(width=0.2))+
  geom_hline(aes(yintercept=median(mean_coverage)),linewidth=1, linetype = 'dotted', col = 'red')+
  scale_color_viridis_d() +
  theme_minimal()+
  theme(axis.text.x=element_text(angle=45)) +
  labs(x="%GC bin", y="Mean coverage per bin")

png("r_analysis/twist_gc_bias_combined.png", width = 15, 
    height = 10, units = "in", res = 300)
 cowplot::plot_grid(gc_bin_num_regions,gc_bin_log10_width, 
                    gc_bin_mean_coverage_plot, gc_bin_length_coverage_plot,
                    gc_bin_norm_coverage_plot,nrow=2,labels="AUTO")
dev.off()
```
